src/Lean/Elab/Tactic/BVDecide/Frontend/Normalize/AC.lean

```diff
@@ -1,39 +1,408 @@
+  instantiateMVars proof
+
+
+/--
+Given an expression `P lhs rhs`, where `lhs, rhs : ty` and `P : $ty → $ty → _`,
+canonicalize top-level applications of a recognized associative and commutative
+operation on both the `lhs` and the `rhs` such that the final expression is:
+  `P ($common ⊕ $lhs') ($common ⊕ $rhs')`
+That is, in a way that exposes terms that are shared between the lhs and rhs.
+
+For example, `x₁ * (y₁ * z) == x₂ * (y₂ * z)` is normalized to
+`z * (x₁ * y₁) == z * (x₂ * y₂)`, pulling the shared variable `z` to the front on
+both sides.
+
+Note that if both lhs and rhs are applications of a *different* operation, we
+canonicalize according to the *left* operation, meaning we treat the entire rhs
+as an atom. This is still useful, as it will pull out an occurence of the rhs
+in the lhs (if present) to the front (such an occurence would be the common
+expression). For example `x + y + ((x * y) + x) = x * y` will be canonicalized
+to `(x * y) + ... = x * y`.
+
+See `Op.fromExpr?` to see which operations are recognized.
+Other operations are ignored, even if they are associative and commutative.
+-/
+def canonicalizeWithSharing (P : Expr) (lhs rhs : Expr) : SimpM Simp.Step := do
+  withTraceNode (collapsed := false) `Meta.AC (fun _ => pure m!"canonicalizeWithSharing") <| do
+  trace[Meta.AC] "Canonicalizing: {indentExpr <| mkApp2 P lhs rhs}"
+
+  let some op := Op.ofApp2? lhs |
+    trace[Meta.AC] "Failed to recognize operation: {indentExpr lhs}"
+    return .continue
+  let some op' := Op.ofApp2? rhs |
+    trace[Meta.AC] "Failed to recognize operation: {indentExpr rhs}"
+    return .continue
+
+  -- Ignore cases where LHS and RHS ops are different.
+  if op != op' then
+    trace[Meta.AC] "Operations mismatch:\
+      the left-hand-side has operation {op}
+        {indentExpr lhs}
+      but the right-hand-side has operation {op'}
+        {indentExpr rhs}"
+    return .continue
+
+  trace[Meta.AC] "Canonicalizing with respect to operation: '{op}' K"
+
+  VarStateM.run' (s:= { op }) do
+    let lCoeff ← computeCoefficients op lhs
+    let rCoeff ← computeCoefficients op rhs
+
+    let ⟨commonCoeff, lCoeff, rCoeff⟩ ← SharedCoefficients.compute lCoeff rCoeff
+    let commonExpr? : Option Expr ← commonCoeff.toExpr op
+    let lNew? : Option Expr ← lCoeff.toExpr op
+    let rNew? : Option Expr ← rCoeff.toExpr op
+
+    -- Since `lCoeff_{old} = commonCoeff + lCoeff_{new}`, and all coefficients
+    -- of `lCoeff_{old}` are zero iff `lExpr` contains only neutral elements,
+    -- we default to `lNew` being some canonical neutral element if both
+    -- `commonExpr?` and `lNew?` are `none`.
+    let lNew := Option.merge (mkApp2 op.toExpr) commonExpr? lNew? |>.getD op.neutralElement
+    let rNew := Option.merge (mkApp2 op.toExpr) commonExpr? rNew? |>.getD op.neutralElement
+
+    let oldExpr := mkApp2 P lhs rhs
+    let expr := mkApp2 P lNew rNew
+    check oldExpr
+    check expr
```

Comment:
Is there ever a situation where we expect this not to type check? Recall that we are type checking potentially ginormous terms here.